pkg/parser/pdf/parser_metadata.go

```diff
@@ -0,0 +1,207 @@
+package pdf
+
+import (
+	"fmt"
+	"io"
+	"strconv"
+	"strings"
+
+	"github.com/google/uuid"
+	"github.com/pdfcpu/pdfcpu/pkg/pdfcpu"
+	"github.com/pkg/errors"
+	"github.com/readium/go-toolkit/pkg/internal/extensions"
+	"github.com/readium/go-toolkit/pkg/manifest"
+	"github.com/readium/go-toolkit/pkg/mediatype"
+	"github.com/trimmer-io/go-xmp/xmp"
+)
+
+// This is completely random
+var UUIDNameSpaceForPDF = uuid.Must(uuid.Parse("4a706cb0-458c-4180-9601-086121ee8d9f"))
+
+func loadDecoder(meta pdfcpu.Metadata) (*xmp.Document, []byte, error) {
+	metabin, err := io.ReadAll(meta.Reader)
+	if err != nil {
+		return nil, nil, errors.Wrap(err, "failed reading XMP metadata block")
+	}
+	doc := &xmp.Document{}
+	if err := xmp.Unmarshal(metabin, doc); err != nil {
+		return nil, nil, errors.Wrap(err, "failed decoding XMP metadata")
+	}
+	return doc, metabin, nil
+}
+
+func ParseMetadata(ctx *pdfcpu.Context, link *manifest.Link) (m manifest.Manifest, err error) {
+	if link != nil {
+		m.ReadingOrder = manifest.LinkList{{
+			Href:       strings.TrimPrefix(link.Href, "/"),
+			Type:       mediatype.PDF.String(),
+			Title:      link.Title,
+			Rels:       link.Rels,
+			Properties: link.Properties,
+			Alternates: link.Alternates,
+			Children:   link.Children,
+			Languages:  link.Languages,
+		}}
+	}
+	m.Metadata.ConformsTo = manifest.Profiles{manifest.ProfilePDF}
+
+	hashmaterial := make([]string, 0, 64)
+	metas, _ := ctx.ExtractMetadata()
+	for _, meta := range metas {
+		doc, metabin, derr := loadDecoder(meta)
+		if derr != nil {
+			err = derr
+			return
+		}
+		defer doc.Close()
+		hashmaterial = append(hashmaterial, string(metabin))
+
+		// b, _ := json.MarshalIndent(doc, "", "  ")
+		// println(string(b))
+
+		err = ParseXMPMetadata(doc, &m.Metadata)
+		if err != nil {
+			return
+		}
+	}
+
+	err = ParsePDFMetadata(ctx, &m)
+	if err != nil {
+		return
+	}
+
+	if m.Metadata.Identifier == "" {
+		// Create a UUIDv5 which is based on a SHA1 hash of the PDF's metadata.
```

Comment:
We used to generate a MD5 hash of the source file on the mobile toolkits (and maybe still do in some cases), but we're moving away from this and just use a `null` identifier if there's no intrinsic identifier in a publication format (or we can't get it).

For PDF, the specification actually mentions the `ID` array.

> A PDF file identifier is defined in the PDF specification as an array of two strings, the first of which is a permanent identifier that doesn’t change even when the file is updated. The second string changes each time the file is updated. For more information, see PDF Reference: Version 1.3 (Second Edition), Adobe Systems Incorporated.

Which I think you can get in `pdfcpu` with this: https://github.com/pdfcpu/pdfcpu/blob/07d97625e3fa00f05a6f7559a53463574329931a/pkg/pdfcpu/xreftable.go#LL101C3-L101C3

On the Swift toolkit, I used only the first identifier in the array (the stable one) and ignored the second one (updated every time the PDF is remade), and converted the bytes into an hexadecimal string:

https://github.com/readium/swift-toolkit/blob/e1735017b7c46ac62599e63adde961367a987b6f/Sources/Shared/Toolkit/PDF/CGPDF.swift#L20-L36

Reply:
Thanks for the pointer to the PDF ID, I had forgotten about that because I was too focused on the xmp metadata. I've switched to using that first if it's there.
I definitely don't want to hash the file, because that's very inefficient in streaming use cases. From my testing, sometimes the identifier is already a hex string (like `88EC969638074476AC5DDC6638B9BB5E`) but can also be bytes in others, so I copied the hex conversion. Will also keep identifier a bit more consistent if syncing of data between toolkits is necessary.